evm/vm/base.py

```diff
@@ -90,13 +121,62 @@ def logger(self):
     #
     # Execution
     #
+    def add_transaction(self, transaction, computation):
+        """
+        Add a transaction to the given block.
+        """
+        receipt = self.make_receipt(transaction, computation)
+
+        transaction_idx = len(self.block.transactions)
+
+        index_key = rlp.encode(transaction_idx, sedes=rlp.sedes.big_endian_int)
+
+        self.block.transactions.append(transaction)
+
+        tx_root_hash = self.block.chaindb.add_transaction(self.block.header, index_key, transaction)
+        receipt_root_hash = self.block.chaindb.add_receipt(self.block.header, index_key, receipt)
+
+        self.block.bloom_filter |= receipt.bloom
+
+        self.block.header.transaction_root = tx_root_hash
+        self.block.header.receipt_root = receipt_root_hash
+        self.block.header.bloom = int(self.block.bloom_filter)
+        self.block.header.gas_used = receipt.gas_used
+
+        # Return `self.block`, not `block`
+        return self.block
+
     def apply_transaction(self, transaction):
         """
         Apply the transaction to the vm in the current block.
         """
         computation = self.execute_transaction(transaction)
         self.clear_journal()
-        self.block.add_transaction(transaction, computation)
+
+        self.block = self.add_transaction(
+            transaction,
+            computation,
+        )
+
+        return computation
+
+    def apply_transaction_to_block(self, transaction, block, chaindb):
+        """
+        Apply the transaction to the vm in the current block.
+        """
+        # (semi) pure function setting
+        self.chaindb = copy.deepcopy(chaindb)
```

Comment:
I suspect the use of `deepcopy` here is not going to be a workable solution since we almost certainly cannot write our `chaindb` class in a way that allows for arbitrary copying of the database.  Solutions that come to mind are:

- Something like the `OverlayDB` class.
- An API on the `BaseChainDB` class such as `chaindb.get_overlay_copy()` which wrapps the underlying db in an `OverlayDB`.